File: src/actions_py/actions_py/robot_move_server.py
#!/usr/bin/env python3 
import rclpy 
from rclpy.node import Node 
import time
import threading

from rclpy.action.server import ServerGoalHandle
from rclpy.action import ActionServer, GoalResponse, CancelResponse
from my_robot_interfaces.action import MoveRobot

from rclpy.executors import MultiThreadedExecutor
from rclpy.callback_groups import ReentrantCallbackGroup


class RobotMoveServerNode(Node):
    def __init__(self):
        super().__init__("robot_move_server")
        self.robot_move_server_ = ActionServer(self, MoveRobot, "robot_move", 
                                               goal_callback= self.goal_callback,
                                               handle_accepted_callback= self.handle_callback, 
                                               cancel_callback=self.cancel_callback,
                                               execute_callback=self.execute_callback,
                                               callback_group=ReentrantCallbackGroup())
        self.lock_ = threading.Lock()
        self.goal_handle_ = None
        self.robot_position = 50
        self.get_logger().info("Server up")

    def goal_callback(self, goal_request: MoveRobot.Goal):
        if goal_request.velocity > 0:
            return GoalResponse.ACCEPT
        else:
            return GoalResponse.REJECT

    def handle_callback(self, new_goal_handle: ServerGoalHandle):
        with self.lock_:
            old_goal_handle = self.goal_handle_
            if old_goal_handle is not None and old_goal_handle.is_active:
                old_goal_handle.abort()
                self.get_logger().info("Abort previous job")
            self.goal_handle_ = new_goal_handle
        self.goal_handle_.execute()

    def cancel_callback(self, goal_handle: ServerGoalHandle):
        self.get_logger().info("Goal is canceling...")
        return CancelResponse.ACCEPT

    def execute_callback(self, goal_handle: ServerGoalHandle):
        self.get_logger().info("Start executing !")
        self.goal_handle_ = goal_handle
        position = goal_handle.request.position
        velocity = goal_handle.request.velocity
        feedback = MoveRobot.Feedback()
        result = MoveRobot.Result()
        self.get_logger().info("Receive goal !")
        while(position > self.robot_position):
            self.get_logger().info("Target position : " + str(position) + ", Robot position : " + str(self.robot_position))
            if not goal_handle.is_active:
                result.position = self.robot_position
                result.message = "Abort the job !"
                return result
            if goal_handle.is_cancel_requested:
                result.position = self.robot_position
                goal_handle.canceled()
                result.message = "The job is canceled!"
                return result
            feedback.current_position = self.robot_position
            goal_handle.publish_feedback(feedback)
            distance = position - self.robot_position
            if distance >= velocity:
                self.robot_position += velocity
            elif distance < velocity:
                self.robot_position += distance
            time.sleep(1.0)
        while(position < self.robot_position):
            self.get_logger().info("Target position : " + str(position) + ", Robot position : " + str(self.robot_position))
            if not goal_handle.is_active:
                result.position = self.robot_position
                result.message = "Abort the job !"
                return result
            if goal_handle.is_cancel_requested:
                result.position = self.robot_position
                goal_handle.canceled()
                result.message = "The job is canceled!"
                return result
            feedback.current_position = self.robot_position
            goal_handle.publish_feedback(feedback)
            distance = self.robot_position -position
            if distance >= velocity:
                    self.robot_position -= velocity
            elif distance < velocity:
                    self.robot_position -= distance
            time.sleep(1.0)
        self.get_logger().info("Target position : " + str(position) + ", Robot position : " + str(self.robot_position))
        feedback.current_position = self.robot_position
        goal_handle.publish_feedback(feedback)
        if self.robot_position == position:
            goal_handle.succeed()
            result.position = self.robot_position
            result.message = "Reach targe position !"
            self.get_logger().info("Reach target position !")
            return result
                

        

def main(args=None): 
    rclpy.init(args=args) 
    node = RobotMoveServerNode()  
    rclpy.spin(node,MultiThreadedExecutor())    
    rclpy.shutdown() 

if __name__ == "__main__":  
    main()

File: src/actions_cpp/src/robot_move_server.cpp
#include "rclcpp/rclcpp.hpp"
#include "rclcpp_action/rclcpp_action.hpp"
#include "my_robot_interfaces/action/move_robot.hpp"

using MoveRobot = my_robot_interfaces::action::MoveRobot;
using namespace std::placeholders;

class RobotMoveServerNode : public rclcpp::Node
{
public:
    RobotMoveServerNode() : Node("robot_move_server"), current_position(50)
    {
        cb_group_ = this->create_callback_group(rclcpp::CallbackGroupType::Reentrant);
        server_ = rclcpp_action::create_server<MoveRobot>(this, "robot_move",
                std::bind(&RobotMoveServerNode::goal_callback, this, _1, _2),
                std::bind(&RobotMoveServerNode::cancel_callback, this, _1),
                std::bind(&RobotMoveServerNode::handle_accept_callback, this, _1),
                rcl_action_server_get_default_options(),
                cb_group_
                );
        RCLCPP_INFO(this->get_logger(), "Robot Move Server is on !");

    }
private:
    
    rclcpp_action::GoalResponse goal_callback(const rclcpp_action::GoalUUID &uuid, std::shared_ptr<const MoveRobot::Goal> goal)
    {
        (void)uuid;
        {
            std::lock_guard<std::mutex> lock(mutex_);
            if (goal_handle_){
                if (goal_handle_->is_active()){
                    preempt_goal_UUID = goal_handle_->get_goal_id();
                }
            }
        }
        if ((goal->position <= 100 && goal->position >= 0) && goal->velocity > 0){
            RCLCPP_INFO(this->get_logger(), "Goal is accepted !");
            return rclcpp_action::GoalResponse::ACCEPT_AND_EXECUTE;
        }
        

        RCLCPP_INFO(this->get_logger(), "Goal is rejected !");
        return rclcpp_action::GoalResponse::REJECT;
        
    }

    rclcpp_action::CancelResponse cancel_callback(const std::shared_ptr<rclcpp_action::ServerGoalHandle<MoveRobot>> goal_handle)
    {
        (void)goal_handle;
        RCLCPP_INFO(this->get_logger(), "Cancel request is accepted !");
        return rclcpp_action::CancelResponse::ACCEPT;
    }

    void handle_accept_callback(const std::shared_ptr<rclcpp_action::ServerGoalHandle<MoveRobot>> goal_handle)
    {
        RCLCPP_INFO(this->get_logger(), "Executing the goal...");
        execute_callback(goal_handle);
    }

    void execute_callback(const std::shared_ptr<rclcpp_action::ServerGoalHandle<MoveRobot>> goal_handle)
    {
        {
            std::lock_guard<std::mutex> lock(mutex_);
            goal_handle_ = goal_handle;
        }
        
        auto result = std::make_shared<MoveRobot::Result>();
        auto feed_back = std::make_shared<MoveRobot::Feedback>();
        auto position = goal_handle->get_goal()->position;
        auto velocity = goal_handle->get_goal()->velocity;
        RCLCPP_INFO(this->get_logger(), "Target: %d", (int)position);
        rclcpp::Rate loop_rate(1);
        while (position != current_position)
        {
            distance = std::abs(position - current_position);
            
            {
                std::lock_guard<std::mutex> lock(mutex_);
                if (goal_handle->get_goal_id() == preempt_goal_UUID){
                    result->position = current_position; 
                    result->message = std::string("Aborted !");   
                    goal_handle->abort(result);
                    return;
                }
            }
            if (goal_handle->is_canceling()){
                result->position = current_position; 
                result->message = std::string("Canceled !");   
                goal_handle->canceled(result);
                return;
            }

            if (position > current_position){
                if (distance < velocity){
                    current_position += distance;
                }else{
                    current_position += velocity;
                }
            }
            if (position < current_position){
                if (distance < velocity){
                    current_position -= distance;
                }else{
                    current_position -= velocity;
                }
            }
            feed_back->current_position = current_position;
            goal_handle->publish_feedback(feed_back);
            RCLCPP_INFO(this->get_logger(), "%d", current_position);
            loop_rate.sleep();
        }
        result->position = current_position; 
        result->message = std::string("Reach target postion !");   
        goal_handle->succeed(result);
        
    }

    rclcpp_action::Server<MoveRobot>::SharedPtr server_;
    int current_position, distance;
    std::mutex mutex_;
    std::shared_ptr<rclcpp_action::ServerGoalHandle<MoveRobot>> goal_handle_;
    rclcpp_action::GoalUUID preempt_goal_UUID;
    rclcpp::CallbackGroup::SharedPtr cb_group_;
};

int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<RobotMoveServerNode>();
    rclcpp::executors::MultiThreadedExecutor executor;
    executor.add_node(node);
    executor.spin();
    rclcpp::shutdown();
    return 0;
}